### ISSUE TITLE
[RFC] S3 Checkpointing Support

### ISSUE THREAD

**Issue by ralovets** [verl-project/verl#7641]
verl supports checkpoint save and resume through filesystem paths today. For FSDP2, each rank writes its sharded model, optimizer, scheduler, and RNG state to the filesystem, while the trainer saves dataloader state and tracks the latest local checkpoint. What Verl does not provide is a built-in object-storage checkpoint backend: there is no distributed S3 writer, remote checkpoint discovery, or remote restore path. A correct staged-upload implementation would still require per-rank uploads, retry and completion tracking, atomic step commits, and remote resume. This proposal adds that lifecycle directly to FSDP2 and overlaps serialization with upload to reduce the time until a checkpoint is durable in S3.

## Public API

Add one trainer setting:

```yaml
trainer:
  default_local_dir: /checkpoints
  default_remote_dir: s3://my-bucket/training/run-123
```

The trainer derives the step path and passes it as `remote_path`. The path uses an `s3://` URI. The trainer owns checkpoint lifecycle, the training backend owns the format, and the S3 adapter owns remote I/O and retries.

## Proposed FSDP2 object-storage save flow

This table describes the proposed object-storage flow, which would use PyTorch Distributed Checkpoint to save model and optimizer state:

| Stage | Proposed behavior |
| --- | --- |
| GPU snapshot | Each rank copies its local model and optimizer state from GPU to CPU memory. |
| Serialize and upload | Each rank serializes its snapshot and uploads completed chunks without waiting for the entire checkpoint file. |
| Shared state | Rank 0 stores the training step, dataloader position, scheduler state, and optimizer counters in `actor/common.pt`. |
| Rank-specific state | Each rank saves its RNG state and, when applicable, its data-sampler state alongside its shard. |
| Actor checkpoint | The initial release would checkpoint the actor only. Its tensor shards and shared state would be grouped under `actor/`. |
| Actor complete | The writer creates `actor/.metadata` after every actor shard and `common.pt` uploaded successfully. |
| Step complete | Rank 0 writes `_COMMITTED`, marking the step safe to restore and recording the step number and number of ranks. |

Using S3, a checkpoint looks roughly like this:

```text
s3://my-bucket/training/run-123/global_step_100/
├── actor/
│   ├── __0_0.distcp
│   ├── __1_0.distcp
│   ├── ...
│   ├── common.pt
│   └── .metadata
└── _COMMITTED
```

Training may continue once the CPU snapshot finishes, but the checkpoint is recoverable only after the actor checkpoint finishes and `_COMMITTED` appears. The checkpoint manager tracks background work. If an upload fails, the trainer raises the error before the next checkpoint or during shutdown. Fail-fast mode can stop training immediately, while a clean shutdown waits for pending uploads to finish.

## Resume and failure handling

Rank 0 selects the newest `_COMMITTED` step, verifies `actor/.metadata`, and broadcasts the selected step. Each rank reads its tensor and RNG state plus shared `common.pt`. The first release requires the same FSDP world size. Missing, corrupt, or incompatible data stops restore before the live model changes. The commit marker keeps failure behavior simple:

| Situation | Result |
| --- | --- |
| Save fails or the job stops before `_COMMITTED` | The incomplete checkpoint is ignored; resume uses the previous one |
| Clean shutdown starts during an upload | Shutdown waits and reports any upload failure |
| `_COMMITTED` is written | The new checkpoint is complete and available for resume |

The previous committed checkpoint is not deleted until a newer checkpoint commits successfully.

The S3 client limits concurrency per rank and retries transient failures with backoff and jitter. Permission and checksum failures are fatal. A retry replaces or verifies expected objects; existence alone is not treated as success.

## Performance optimizations

| Optimization | Effect |
| --- | --- |
| Replicated state placement | Replicated state placement stores small optimizer state shared across all FSDP ranks, including per-parameter optimizer step counters and parameter-group settings such as learning rate, betas, epsilon, weight decay, and parameter names. By default, DCP deduplicates this state and stores each value in only one rank’s shard. This is inefficient for object storage because other ranks may need to read a large remote shard to recover a few small values. We support storing the shared state once in `common.pt` or duplicating it into every rank’s shard, allowing each rank to restore without reading another rank’s shard. |
| NUMA-aware workers | Uses CPU memory attached to the nearest CPU socket, reducing GPU-to-CPU transfer time. |
| Reusable pinned CPU buffers | Allocates staging memory once and reuses it. GPU-to-CPU copies run asynchronously, with one wait after all copies are scheduled. |
| Cached DCP planning | Reuses the distributed save plan after verifying that the tensor layout is unchanged. |
| Streaming S3 uploads | Uploads completed 64 MiB parts while DCP serializes the rest of the shard. The larger parts reduce request volume and avoid S3 throttling. |
| Cached finalization | Reuses the tensor-location index instead of gathering hundreds of thousands of records for every save. Layout changes fall back to the full DCP path. |
| Direct `.metadata` publication | Uploads `actor/.metadata` directly to its final S3 key, avoiding a temporary upload followed by copy and delete. `_COMMITTED` is written later, after the whole step is complete. |

## Backend support

| Priority | Backend | Plan |
| --- | --- | --- |
| Initial | FSDP2 | S3 checkpoint save and restore |
| Future | Other backends | Integrate through their existing checkpoint implementations |

The initial implementation focuses on FSDP2. Other backends can add S3 support without adopting the FSDP2 checkpoint format.

## Results and release bar

The design was tested with Qwen3.5-122B-A10B on 8 `p5.48xlarge` AWS instances and 64 FSDP ranks, using pinned CPU staging, NUMA binding, plan caching, and streaming upload.

| Step | Timing |
| --- | ---: |
| `save_state` request, including 3.8s staging | **5.1s** |
| Async background save | **34.4s** |
| Total durable save | **39.4s** |
| Clean load | **78.1s** |

The checkpoint was about 1.37 TiB. Training resumed after `save_state` returned, while the background save continued until the checkpoint became durable in S3.

Before merge, an end-to-end test must save, restart, restore, and take another optimizer step equivalent to uninterrupted training. The workflow should actually run the S3 test.

## Implementation plan

| PR | Scope |
| --- | --- |
| 1. Synchronous foundation | Add the S3 client and base modules for URI parsing, object reads and writes, listing, and retries. Add `default_remote_dir`, `remote_path`, synchronous FSDP2 save and restore, `common.pt`, `.metadata`, `_COMMITTED`, validation, and failure tests. |
| 2. Async CPU snapshot | Add NUMA-aware workers, reusable pinned buffers, background error propagation, and clean shutdown handling. |
| 3. DCP caching | Cache DCP planning and finalization data, with layout validation and fallback to the full path. |
| 4. Streaming uploads | Upload 64 MiB parts during serialization, publish `.metadata` directly, bound request concurrency, and add performance metrics. |

The synchronous path remains the correctness baseline and fallback. Each optimization PR must include its own benchmark and failure coverage.

Out of scope: restore with a different FSDP world size, Hugging Face export, and changes to the local checkpoint format.

**Comment by wuxibin89** [comment]
@ralovets Can we mount S3 via FUSE? Then we can write S3 bucket as local filesystem. This is exactly how we save checkpoint to HDFS, we also mount HDFS via FUSE.
